File: src/test/java/org/kowboy/task/LightLevelTaskTest.java
package org.kowboy.task;

import org.bukkit.Bukkit;
import org.bukkit.Server;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitScheduler;
import org.bukkit.scheduler.BukkitTask;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.kowboy.task.LightLevelTask;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.logging.Logger;

import static org.mockito.Mockito.*;

public class LightLevelTaskTest {
    @Mock private JavaPlugin pluginMock;
    @Mock private Player playerMock;
    @Mock private Runnable onCancelMock;
    @Mock private BukkitTask taskMock;
    private static BukkitScheduler schedulerMock;
    private static Logger loggerMock;

    private LightLevelTask task;

    @BeforeClass
    public static void beforeClass() {
        Server serverMock = mock(Server.class);
        loggerMock = mock(Logger.class);
        schedulerMock = mock(BukkitScheduler.class);

        when(serverMock.getScheduler()).thenReturn(schedulerMock);
        when(serverMock.getLogger()).thenReturn(loggerMock);

        Bukkit.setServer(serverMock);
    }

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        when(taskMock.getTaskId()).thenReturn(1);

        task = new LightLevelTask(2, playerMock, onCancelMock);

        reset(schedulerMock);
        when(schedulerMock.runTask(pluginMock, (Runnable) task)).thenReturn(taskMock);
        task.runTask(pluginMock);
    }


    @Test
    public void testPlayerOffline() {
        when(playerMock.isOnline()).thenReturn(false);
        task.run();
        verify(onCancelMock).run();
    }

    @Test
    public void testPlayerNotSneaking() {
        when(playerMock.isOnline()).thenReturn(true);
        when(playerMock.isSneaking()).thenReturn(false);
        task.run();
        verify(onCancelMock).run();
    }
}

File: src/test/java/org/kowboy/task/LightLevelTaskTest.kt
package org.kowboy.task

import io.mockk.*
import io.mockk.impl.annotations.MockK
import org.bukkit.Bukkit
import org.bukkit.Server
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitScheduler
import org.bukkit.scheduler.BukkitTask
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import java.util.logging.Logger

class LightLevelTaskTest {
    @MockK lateinit var playerMock: Player
    @MockK lateinit var pluginMock: JavaPlugin
    @MockK lateinit var onCancelMock: Runnable
    @MockK lateinit var taskMock: BukkitTask

    private lateinit var sut: LightLevelTask

    companion object {
        private lateinit var schedulerMock: BukkitScheduler
        private lateinit var loggerMock: Logger

        @BeforeClass
        @JvmStatic
        fun beforeClass() {
            val serverMock = mockk<Server>(relaxed = true)
            loggerMock = mockk(relaxed = true)
            schedulerMock = mockk(relaxed = true)
            every { serverMock.scheduler } returns schedulerMock
            every { serverMock.logger } returns loggerMock
            every { serverMock.name } returns "Test Server"
            Bukkit.setServer(serverMock)
        }
    }

    @Before
    fun setup() {
        MockKAnnotations.init(this, relaxUnitFun = true)
        every { taskMock.taskId } returns 1
        sut= LightLevelTask(2, playerMock, onCancelMock)

        clearMocks(schedulerMock)
        every { schedulerMock.runTask(pluginMock, sut as Runnable) } returns taskMock
        sut.runTask(pluginMock)
    }

    @Test
    fun testPlayerOffline() {
        every { playerMock.isOnline } returns false
        sut.run()
        verify { onCancelMock.run() }
    }

    @Test
    fun testPlayerNotSneaking() {
        every { playerMock.isOnline } returns true
        every { playerMock.isSneaking } returns false

        sut.run()
        verify { onCancelMock.run() }
    }

}